### ISSUE TITLE
Combining paths in the glyph φ doesn't work.

### ISSUE THREAD

**Issue by matj1** [glyphr-studio/Glyphr-Studio-2#187]
**Describe the bug & what the expected behavior should be:**

I try to combine two shapes (the ring and the stem) into one in the glyph φ, but it says “Combine shapes error: No overlapping shapes.” instead. I expect that it would combine them into one path as the outline of the letter and two smaller paths as the holes.

**Steps to reproduce the bug:**

1. Load [my font project](https://github.com/glyphr-studio/Glyphr-Studio-2/files/14086087/Greek.Serif.gs2.json) without the “.json” in the name.
2. Select the glyph for φ (or θ or maybe something else where it happens)
3. Click the button “Combine all paths” shown as a square overlapping a circle.

**Other notes & screenshots:**

This is how it looks:

![How it looks](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/22873928/c2deb662-93cc-4c2d-ae7a-c6e5847e9c1f)

**Comment by matj1** [comment]
I tried it with a new glyph in a new project, and it was inconsistent; I mean that the combining happened, but it deleted the part of the stem in the hole, which too is incorrect:

![Snímek obrazovky (6)](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/22873928/a4e388a3-85e2-4303-a7c9-20e85cea5983)

And the result:

![Snímek obrazovky (7)](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/22873928/da1b00ab-6077-4597-922c-6aa915c18d8d)

There should be two holes.

**Comment by mattlag** [comment]
This is a good bug - thank you for submitting it. By "good bug" I mean it may be very tricky to fix 🙂. 

If I remember correctly, the combine algorithm fist sorts shapes by winding, and combines them. Actually, I think the algorithm assumes shapes with different winding will never need to be combined. In order for this combine to be successful, the black shapes will have to influence the transparent shapes.

I will think about this - but, there may be no short-term fix. For the short-term, I would recommend exporting the SVG for this glyph and working with it in another graphics program (like Illustrator or Inkscape) then re-importing the merged shapes.

**Comment by matj1** [comment]
One other issue with combining is in the cases with just one shape intersecting itself. I expect that Combine all paths would transform it to a path for the outline and a path for the hole (and maybe other paths for other holes). But the button is not available with just one path. An example of that is β in my font.

**Comment by mattlag** [comment]
Interesting - this is a case I had not considered. I will think about that as well. 

As a work-around, you could draw a shape that 'covers' the self-overlap. Then you have two shapes to select, which enables the combine shapes action - and the 'covering' shape will eliminate the self-overlap region:

Before:
![image](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/5897669/390c8e0f-60de-4617-90d2-4819d896eb1d)
 
After:
![image](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/5897669/32fe30eb-ff97-4e15-938f-8ba0119933c7)

**Comment by matj1** [comment]
I mean rather something like this:

![My self-overlapping shape](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/22873928/24d3d96a-916c-4b00-a5fe-ddce17c62299)

There is also an other shape covering the intersection. The result of combining is very strange. I displayed layers especially to see which shapes are there after the combining.

![My self-overlapping shape after combining.](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/22873928/add06a7a-2a99-45f4-811d-7a54a2a68b8f)

**Comment by matj1** [comment]
Also, I tried something similar to what you showed, and it said “Combining shapes error: No overlapping shapes found with the same winding.”.

![“Combining shapes error: No overlapping shapes found with the same winding.”](https://github.com/glyphr-studio/Glyphr-Studio-2/assets/22873928/b916d8d7-a008-4b7e-a18a-7861400070ec)

**Comment by matj1** [comment]
Also note that I can't make a shape covering the whole self-overlap and not adding anything to the shape. That is because, in my cases, the intersections are on curves, and calculating the exact coordinates of the intersections would be very difficult.

**Comment by mattlag** [comment]
I will try to think about these two issues, and how to update the Combine Shapes algorithm. You have outlined good test cases, so hopefully I can make changes that meet these two cases.

**Comment by mattlag** [comment]
Hello!

So... I went off the deep end fixing this bug. I basically gave up on our internal Bezier Curve overlap algorithm, and just stuck Paper.js into the project to do all our Shape Combine actions. v2.1.4 just shipped, and not only does it have better "Unite" multi-shape action, but there is also now Subtract, Divide, Exclude, and Union.

That's the fix. Unfortunately, I was thinking more about the PHI problem. Two circles and a vertical rectangle will *never* boolean combine in a way that you want. It's impossible. If you try this same thing in Illustrator, you'll get the same result.

The process for PHI will be something like:
1. Make two copies of your vertical rectangle
2. Select one of the vertical rectangles, make sure it's above the small interior circle, and use the rectangle to Subtract from the small interior circle.  This will give you two "D" shapes.
3. With your second vertical rectangle and your large circle, select both and Unite them

There is no way to select all three shapes, do some combine action, and get the result you want.  It has to be split into two separate combine actions.

Thank you for starting me down this trail. The result isn't perfect, but we have new Combine features that are more dependable than before.

**Comment by matj1** [comment]
> Unfortunately, I was thinking more about the PHI problem. Two circles and a vertical rectangle will _never_ boolean combine in a way that you want. It's impossible. If you try this same thing in Illustrator, you'll get the same result.

I tried to combine it in Inkscape and in FontForge, and it behaved how I wanted in these cases.

**Comment by mattlag** [comment]
You are correct! Although I have to be careful, Illustrator and Inkscape require an additional step of creating a "Compound Shape" in order to create the cut-out. There is no such concept in fonts.

But, FontForge does offer a "Remove Overlaps" action, which works as intended. It must somehow take into consideration the winding of non-intersecting shapes. Currently as I have it implemented, I just use the out-of-the-box Paper.js "Unite" function.

I'm going to re-open this, because there will have to be some additional work to move from "Unite" to "Remove Overlaps".

**Comment by mattlag** [comment]
Fixed in 2.9.0
